### PR TITLE
[Collection] Refactor

### PR DESCRIPTION
**Pull request by rozPierog**
I should really change [this line](https://github.com/Nozbe/WatermelonDB/blob/collection-refactor/src/observation/simpleObserver/index.js#L30) to something different but I can't find other way to do it.

This PR prepares `Collection` and `simpleObserver` for getting changes in an array and also adds `CollectionChangeTypes` for differentiation between different changes in Collection and not only `isDestroyed: boolean`

### CODE REVIEW

File: src/Database/index.js [hunk 1]
```diff
@@ -66,7 +66,7 @@ export default class Database {
   }
 
   // Emits a signal immediately, and on change in any of the passed tables
-  withChangesForTables(tables: TableName<any>[]): Observable<CollectionChange<any> | null> {
+  withChangesForTables(tables: TableName<any>[]): Observable<CollectionChangeSet<any> | null> {
```
[the remark this answers is not in the document]
Comment (rozPierog) [reply src/Database/index.js]: It's also used in `obseveCount` and `reloadingObserver`. Demo app works fine, let me know if you want test coverage for those methods.

File: src/observation/simpleObserver/index.js [hunk 2]
```diff
@@ -25,29 +26,33 @@ function observeChanges<Record: Model>(query: Query<Record>): (Record[]) => Obse
...
+            // Remove if doesn't match anymore
+            matchingRecords.splice(index, 1)
+            shouldUpdate = true
+          } else if (matches && !currentlyMatching) {
+            // Add if should be included but isn't
+            matchingRecords.push(record)
+            shouldUpdate = true
+          }
+        })
+        if (shouldUpdate) emit()
```
[the remark this answers is not in the document]
Comment (radex) [reply src/observation/simpleObserver/index.js]: @rozPierog what about this? Can you check eslint settings to make sure this is reported in the future?

File: src/Database/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (radex) [reply src/Database/index.js]: Not in this PR, but add it to the backlog — I'm surprised those _don't_ have tests 😱

File: src/observation/simpleObserver/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rozPierog) [reply src/observation/simpleObserver/index.js]: I've added curly rule to eslint 😉

**Comment by radex** [comment]
@rozPierog can you propose the same ESLint change to purple repo?

**Comment by radex** [comment]
Thanks @rozPierog 🎉

**Pull request merged**